### PR TITLE
Add support for libdeflate

### PR DESCRIPTION
**Pull request by oliv3r**
Libdeflate is an advanced compression algorithm which is 100% gzip (binary) compatible.

This is a re-do of https://github.com/openwrt/openwrt/pull/10336 so see all previous comments there.

The major difference is that this Merge Request only adds the option to use deflate, rather then changing the default. This allows targets to use this.

Once proven stable and reliable on a few targets, we can offer a 'generic' option in menuconfig, allowing us to switch the default. Also, I have no idea how to do that ;)

This Merge Request builds ontop of https://github.com/openwrt/openwrt/pull/11506 and to avoid merge conflicts those patches are here too, so lets merge that one first ;)

Alternatively, merge this one and close the other one :)

### CODE REVIEW

**Comment by neheb** [comment]
```
+ curl -f --connect-timeout 20 --retry 5 --location https://sources.cdn.openwrt.org/libdeflate-1.15.tar.xz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404 
Download failed.
+ curl -f --connect-timeout 20 --retry 5 --location https://sources.openwrt.org/libdeflate-1.15.tar.xz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404 Not Found
Download failed.
+ curl -f --connect-timeout 20 --retry 5 --location https://mirror2.openwrt.org/sources/libdeflate-1.15.tar.xz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404 Not Found
Download failed.
No more mirrors to try - giving up.
bash: cmake: command not found
make[2]: *** [Makefile:36: /__w/openwrt/openwrt/openwrt/build_dir/host/libdeflate-1.15/.configured] Error 127
time: tools/libdeflate/compile#0.55#0.15#2.06
```

Sounds like you're missing a dependency on tools/cmake

File: tools/libdeflate/Makefile [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+PKG_SOURCE_URL:=https://github.com/ebiggers/libdeflate.git
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_VERSION:=020133854ff73b8506fe59f92a9b5b622d360716
+PKG_MIRROR_HASH:=122feff4543541b547dc89e832adf262c81911ae1acbccdc591f0353a85b600a
+
+include $(INCLUDE_DIR)/host-build.mk
+include $(INCLUDE_DIR)/cmake.mk
+
+CMAKE_HOST_OPTIONS += \
```
Comment (neheb) [inline tools/libdeflate/Makefile:21]: Probably makes more sense to build static.

Comment (neheb) [reply tools/libdeflate/Makefile:21]: wait this is already default.

File: tools/libdeflate/Makefile [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+# Copyright (C) 2022 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=libdeflate
+PKG_VERSION:=1.15
+PKG_RELEASE:=$(AUTORELEASE)
```
Comment (neheb) [inline tools/libdeflate/Makefile:11]: This is deprecated.

File: tools/libdeflate/Makefile [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+PKG_SOURCE_URL:=https://github.com/ebiggers/libdeflate.git
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_VERSION:=020133854ff73b8506fe59f92a9b5b622d360716
```
Comment (neheb) [inline tools/libdeflate/Makefile:15]: You can replace this with v1.15

**Comment by neheb** [comment]
https://github.com/neheb/openwrt/commit/c80dcaa44569aa5db9b48777082dd7f3baae503c

File: tools/libdeflate/Makefile (hunk 1 above)
Comment (oliv3r) [reply tools/libdeflate/Makefile:21]: It seems to work 'just fine' here obviously the .tar.xz isn't mirrored by openwrt yet? but github only offers a .tar.gz, though openwrt may be smart tod ownload the file automagically as tar.xz?/

As for cmake; I just took the hint from a bunch of others from the 'tools' Makefiles, e.g. firmware-utils, all they do is `include $(INCLUDE_DIR)/cmake.mk` so figured that's enough to satisfy the dependency?

File: tools/libdeflate/Makefile (hunk 3 above)
Comment (oliv3r) [reply tools/libdeflate/Makefile:15]: Can I replace this also with `$(PKG_VERSION)` not duplicate that bit twice?

Comment (neheb) [reply tools/libdeflate/Makefile:15]: Yes

File: tools/libdeflate/Makefile (hunk 1 above)
Comment (neheb) [reply tools/libdeflate/Makefile:21]: Not for tools

Comment (oliv3r) [reply tools/libdeflate/Makefile:21]: so why is 'firmware-utls' under tools? and how do I add the dependency? :)

Comment (neheb) [reply tools/libdeflate/Makefile:21]: I linked to a commit that shows how.

**Comment by oliv3r** [comment]
> [neheb@c80dcaa](https://github.com/neheb/openwrt/commit/c80dcaa44569aa5db9b48777082dd7f3baae503c)

I thought this was a 'mention' :) my bad. Thanks for that!

File: tools/libdeflate/Makefile [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+# Copyright (C) 2022 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=libdeflate
+PKG_VERSION:=1.15
+PKG_RELEASE:=0
```
Comment (neheb) [inline tools/libdeflate/Makefile:11]: Starts with 1

File: tools/libdeflate/Makefile [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_VERSION:=v$(PKG_VERSION)
+PKG_MIRROR_HASH:=122feff4543541b547dc89e832adf262c81911ae1acbccdc591f0353a85b600a
+
+include $(INCLUDE_DIR)/host-build.mk
+include $(INCLUDE_DIR)/cmake.mk
+
+define Host/Install
+	$(INSTALL_BIN) $(HOST_BUILD_DIR)/programs/libdeflate-gzip $(STAGING_DIR_HOST)/bin/
+	ln -f -s libdeflate-gzip $(STAGING_DIR_HOST)/bin/libdeflate-gunzip
```
Comment (neheb) [inline tools/libdeflate/Makefile:23]: I think $(LN) can be used here.

**Comment by neheb** [comment]
I can already tell what @ynezz is going to say about this not being used.

Maybe add a commit that replaces gunzip with libdeflate-gunzip. That should be a safe one.

Is libdeflate faster when decompressing?

**Comment by oliv3r** [comment]
> I can already tell what @ynezz is going to say about this not being used.
> 
> Maybe add a commit that replaces gunzip with libdeflate-gunzip. That should be a safe one.
> 
> Is libdeflate faster when decompressing?

Yes it is, and in the original MR, it was said to NOT replace gzip. I've replaced it on the realtek targets, but I think maintainers should replace and test it on their targets.

**Comment by oliv3r** [comment]
> I can already tell what @ynezz is going to say about this not being used.
> 
> Maybe add a commit that replaces gunzip with libdeflate-gunzip. That should be a safe one.
> 
> Is libdeflate faster when decompressing?

So I've added one case to justify the addition :) Though I don't really agree with it; offer the tool, so we can start using it I'd figure ... :) So kind of a political hack :D

**Comment by robimarko** [comment]
I like this as so far for me it worked on multiple targets just fine, looks to be a bit faster to compress and produces smaller images, so I would say win-win.

`Reviewed-by: Robert Marko <robimarko@gmail.com>`

**Comment by oliv3r** [comment]
@neheb Better?

**Review by neheb — APPROVED** [review]

**Comment by oliv3r** [comment]
> I like this as so far for me it worked on multiple targets just fine, looks to be a bit faster to compress and produces smaller images, so I would say win-win.
> 
> `Reviewed-by: Robert Marko <robimarko@gmail.com>`

So as soon this gets merged, hopefully you can upgrade your targets that you support. I'll hope to get some help on the realtek stuff, as I don't own any of those switches with gzip only support ...

@Ansuel any thoughts?

**Comment by oliv3r** [comment]
@Ansuel rebased and pushed to remove merge conflicts.

**Comment by oliv3r** [comment]
Rebased, and used our realtek stuff instead of the less ideal mediatek one.

@svanheule do you have any of those affected switches and can test it? Seems to work fine on my switch, though I'd rather keep that at lzma ;)

File: target/linux/realtek/image/Makefile [hunk 6]
```diff
@@ -90,12 +90,12 @@ define Device/Default
   KERNEL := \
 	kernel-bin | \
 	append-dtb | \
-	gzip | \
+	libdeflate | \
```
Comment (svanheule) [inline target/linux/realtek/image/Makefile:93]: I suspect people are more likely to understand what a `gzip` command does, compared to `libdeflate`.  Not knowing libdeflate, I was wondering if it only supports gzip, or if it also supports other compression methods. Looking at their README, this looks exactly like the "DEFLATE vs. zlib vs. gzip" ambiguity that is addressed there.

Should this be changed to something more explicit? Maybe `libdeflate-gzip` like the actual command?

**Review by svanheule — COMMENTED** [review]
If you tested gzip-ed uImages, this should be fine even if your bootloaders support lzma too. Since Robert also tested this on other targets, I don't suspect this will cause any issues.

File: target/linux/realtek/image/Makefile (hunk 6 above)
Comment (oliv3r) [reply target/linux/realtek/image/Makefile:93]: Lets do that. While I encourage people learning new things, making things easier to read and understand surely is only beneficial :)

Comment (robimarko) [reply target/linux/realtek/image/Makefile:93]: This is a good idea to make it clear it's compatible and not breaking change.

**Comment by svanheule** [comment]
Thanks for the update. I built a few images to compare with master, would be nice to add this (or your own comparison) to the commit message to support the tooling switch:

```
gzip    | libdeflate-gzip | delta  | image [openwrt-realtek-rtl839x-*]
--------+-----------------+--------+-----------------------------------------------
6589174 | 6298794         | 290380 | d-link_dgs-1210-52-initramfs-kernel.bin
6291632 | 6029488         | 262144 | d-link_dgs-1210-52-squashfs-factory_image1.bin
6292270 | 6030128         | 262142 | d-link_dgs-1210-52-squashfs-sysupgrade.bin
6589142 | 6298760         | 290382 | zyxel_gs1900-48-initramfs-kernel.bin
6292264 | 6030122         | 262142 | zyxel_gs1900-48-squashfs-sysupgrade.bin
```

Don't forget to add Robert's Reviewed-by!

@neheb would like to merge these changes? I can merge them too otherwise, and add a Reviewed-by if you would like to provide one.

**Comment by neheb** [comment]
@svanheule I don't have merge access.

**Comment by neheb** [comment]
Anyway,

```
Reviewed-by: Rosen Penev <rosenp@gmail.com>
```

**Comment by neheb** [comment]
I just ran on the biggest file I could find in dl. Really nice results:

```
mangix@ryzen ~/d/openwrt (mangix) [SIGINT]> time gzip -dc dl/tessdata-4.1.0.tar.gz > /dev/null

________________________________________________________
Executed in    9.09 secs    fish           external
   usr time    8.89 secs    0.00 millis    8.89 secs
   sys time    0.15 secs    1.26 millis    0.15 secs

mangix@ryzen ~/d/openwrt (mangix)> time staging_dir/host/bin/libdeflate-gunzip dl/tessdata-4.1.0.tar.gz > /dev/null

________________________________________________________
Executed in    3.17 secs    fish           external
   usr time    2.40 secs    0.16 millis    2.40 secs
   sys time    0.76 secs    1.03 millis    0.76 secs

****
```

**Comment by oliv3r** [comment]
@svanheule I've updated the commit messages and added the reviewed-by's and some testing data. Should be good to go now.

**Comment by svanheule** [comment]
Thanks everyone!

Merged with:
a51ca085bff2 tools: Add gzip-libdeflate advanced compressor
9260027535b8 realtek: Migrate to libdeflate

**Comment by dhewg** [comment]
I may have missed some comment due to the various PRs and bike-shedding...

...but has anyone tried to use libdeflate for opkg?

That'll be smaller packages without costs (but skewing "Installed-Size" even more, see also https://github.com/openwrt/openwrt/pull/11910#issuecomment-1416721838).

**Comment by neheb** [comment]
Nope. PR needed.

**Pull request closed without merging**